internal/pagination/pagination.go

```diff
@@ -10,14 +10,231 @@ import (
+// to retrieve a hash of a users grants.
+type GrantsHasher interface {
+	GrantsHash(ctx context.Context) ([]byte, error)
+}
+
+// ListResponse represents the response from the paginated list operation.
+type ListResponse[PbT ResponseItem] struct {
+	Items                 []PbT
+	CompleteListing       bool
+	MarshaledRefreshToken string
+	DeletedIds            []string
+	EstimatedItemCount    int
+}
+
+// ListItemsFunc defines the signature of the callback used to list
+// items for a specific resource type.
+type ListItemsFunc[T any] func(prevPageLast T, refreshToken *pbs.ListRefreshToken, limit int) ([]T, error)
+
+// ConvertAndFilterFunc defines the signature of the callback used to
+// convert from T to PbT and filter the results.
+type ConvertAndFilterFunc[T any, PbT ResponseItem] func(item T) (PbT, bool, error)
+
+// PaginateRequest performs refresh token parsing and validation
+// and performs pagination as specified by the request page size
+// and service configured max page size using the itemLister. It
+// will call the itemLister until it has filled a page, or until
+// there are no more results available.
+func PaginateRequest[T any, PbT ResponseItem](
+	ctx context.Context,
+	maxPageSize uint,
+	resourceType pbs.ResourceType,
+	req ListRequest,
+	itemLister ListItemsFunc[T],
+	converterAndFilterer ConvertAndFilterFunc[T, PbT],
+	grantsHasher GrantsHasher,
+	repo Repository,
+) (*ListResponse[PbT], error) {
+	const op = "pagination.PaginateRequest"
+
+	if util.IsNil(grantsHasher) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "grants hasher is required")
+	}
+	if maxPageSize == 0 {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "max page size is required")
+	}
+	if resourceType == pbs.ResourceType_RESOURCE_TYPE_UNSPECIFIED {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "resource type is required")
+	}
+	if util.IsNil(req) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "the request is required")
+	}
+	if itemLister == nil {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "item list function is required")
+	}
+	if converterAndFilterer == nil {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "convert-and-filter function is required")
+	}
+	if util.IsNil(repo) {
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "repository is required")
+	}
+
+	grantsHash, err := grantsHasher.GrantsHash(ctx)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	var refreshToken *pbs.ListRefreshToken
+	if req.GetRefreshToken() != "" {
+		// Note that refresh token parsing and validation happens after authorization,
+		// since validation requires access to the grants hash.
+		refreshToken, err = parseRefreshToken(ctx, req.GetRefreshToken())
+		if err != nil {
+			return nil, errors.Wrap(ctx, err, op)
+		}
+		if err := validateRefreshToken(ctx, refreshToken, grantsHash, resourceType); err != nil {
+			return nil, errors.Wrap(ctx, err, op)
+		}
```

Comment:
Note that parsing and validation of the refresh token now happens _after_ authorization. If we wanted to parse first, we'd have to do authorization in this function, which I wanted to avoid.

Reply:
Is there any reason we wouldn't want to authorise before validating?  Correct me if I'm wrong, but to me it makes more sense to check if the user can even make the request, before checking if the request is correct

Reply:
We generally do authentication -> request validation -> authorization AFAICT. We'll now do authentication -> some request validation -> authorization -> some more request validation.